### DISCUSSION TITLE
Time based rendering

### DISCUSSION THREAD

**Discussion by EMwavelength** [mitsuba-renderer/mitsuba3#210] [Q&A] [answered]
Hi,
  I am sorry if this is a request or an already implemented feature, but i want to know if mitsuba 3 is able to simulate transient effects of light i.e time based rendering both camera and non-camera space . Since mitsuba 3 is research oriented i like to know if such feature is already implemented, if not will it be in roadmap or future cosideration?

**Comment by alhirzel** [comment]
I'm not a core developer but have worked in this area before. Mitsuba supports tracking a `time` for rays as they propagate through a scene:

https://github.com/mitsuba-renderer/mitsuba3/blob/master/include/mitsuba/core/ray.h#L33

Existing integrators seem to treat this only wrt the shutter open/close, for example:

https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/integrators/ptracer.cpp#L171

I suspect that ray length does not couple into this quantity for any of the built-in integrators. You would need to write your own, potentially taking into account the medium properties and disabling the built-in treatment of IOR at the interfaces of heterogeneous media (since it would be redundant with your formulation).

I suspect that this may not have the necessary precision for some applications (e.g. time of flight in the realm of nanoseconds will struggle unless you use double precision arithmetic everywhere).

**Reply by EMwavelength** [reply]
Thank you for the answer.

**Reply by jgarciapueyo** [reply, marked as answer]
Hi, it might be a little bit late for this, but we have developed "mitransient", which is a time-resolved light transport renderer built on top of Mitsuba3. Take a look in case this is what you need: https://github.com/diegoroyo/mitransient
